src/SDL_properties.c

```diff
@@ -0,0 +1,257 @@
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+#include "SDL_internal.h"
+#include "SDL_hashtable.h"
+#include "SDL_properties_c.h"
+
+
+typedef struct
+{
+    void *value;
+    void (SDLCALL *cleanup)(void *userdata, void *value);
+    void *userdata;
+} SDL_Property;
+
+typedef struct
+{
+    SDL_HashTable *props;
+    SDL_Mutex *lock;
+} SDL_Properties;
+
+static SDL_HashTable *SDL_properties;
+static SDL_RWLock *SDL_properties_lock;
+static SDL_PropertiesID SDL_last_properties_id;
+
+
+static void SDL_FreeProperty(const void *key, const void *value, void *data)
+{
+    SDL_Property *property = (SDL_Property *)value;
+    if (property->cleanup) {
+        property->cleanup(property->userdata, property->value);
+    }
+    SDL_free((void *)key);
+    SDL_free((void *)value);
+}
+
+static void SDL_FreeProperties(const void *key, const void *value, void *data)
+{
+    SDL_Properties *properties = (SDL_Properties *)value;
+    if (properties) {
+        if (properties->props) {
+            SDL_DestroyHashTable(properties->props);
+            properties->props = NULL;
+        }
+        if (properties->lock) {
+            SDL_DestroyMutex(properties->lock);
+            properties->lock = NULL;
+        }
+        SDL_free(properties);
+    }
+}
+
+int SDL_InitProperties(void)
+{
+    if (!SDL_properties_lock) {
+        SDL_properties_lock = SDL_CreateRWLock();
+        if (!SDL_properties_lock) {
+            return -1;
+        }
+    }
+    if (!SDL_properties) {
+        SDL_properties = SDL_CreateHashTable(NULL, 16, SDL_HashID, SDL_KeyMatchID, SDL_FreeProperties, SDL_FALSE);
+        if (!SDL_properties) {
+            return -1;
+        }
+    }
+    return 0;
+}
+
+void SDL_QuitProperties(void)
+{
+    if (SDL_properties) {
+        SDL_DestroyHashTable(SDL_properties);
+        SDL_properties = NULL;
+    }
+    if (SDL_properties_lock) {
+        SDL_DestroyRWLock(SDL_properties_lock);
+        SDL_properties_lock = NULL;
+    }
+}
+
+SDL_PropertiesID SDL_CreateProperties(void)
+{
+    SDL_PropertiesID props;
+    SDL_Properties *properties = NULL;
+    SDL_bool inserted = SDL_FALSE;
+
+    properties = SDL_malloc(sizeof(*properties));
+    if (!properties) {
+        goto error;
+    }
+    properties->props = SDL_CreateHashTable(NULL, 4, SDL_HashString, SDL_KeyMatchString, SDL_FreeProperty, SDL_FALSE);
+    if (!properties->props) {
+        goto error;
+    }
+    properties->lock = SDL_CreateMutex();
+    if (!properties->lock) {
+        goto error;
+    }
+
+    if (SDL_InitProperties() < 0) {
+        goto error;
+    }
+
+    if (SDL_LockRWLockForWriting(SDL_properties_lock) == 0) {
+        props = ++SDL_last_properties_id;
+        if (SDL_InsertIntoHashTable(SDL_properties, (const void *)(uintptr_t)props, properties)) {
```

Comment:
Did we land on using uintptr_t?

(If not, I'm fine leaving this alone until some weird platform fails to compile.)

Reply:
props is SDL_PropertiesID which is Uint32 i.e. fixed size, so uintptr_t is needed here, is it not?

Reply:
Yes, but I was wondering if we settled on uintptr_t here, vs size_t, vs adding some sort of SDL_uintptr.

Reply:
In the API we are studiously avoiding the question. In the implementation we actually have to use uintptr_t to avoid compiler warnings on various compilers.